File: 2019/04/solution.py
from collections import Counter


def parse_input():
    with open('input.txt') as f:
        line = f.read()
    start, end = map(int, line.split('-'))
    return list(map(digits, range(start, end + 1)))

def part1(guesses):
    return len([
        guess
        for guess in filter(increasing, guesses)
        if max(Counter(guess).values()) >= 2
    ])

def part2(guesses):
    return len([
        guess
        for guess in filter(increasing, guesses)
        if 2 in Counter(guess).values()
    ])

def digits(guess):
    return [int(c) for c in str(guess)]

def increasing(digits):
    for i in range(len(digits) - 1):
        if digits[i] > digits[i + 1]:
            return False
    return True


guesses = parse_input()
print(part1(guesses))
print(part2(guesses))

File: 2019/04/solution.exs
defmodule AoC2019.Day04 do
  def parse_input() do
    {start, finish} =
      File.stream!("input.txt")
      |> Enum.at(0)
      |> String.split("-")
      |> Enum.map(&String.to_integer/1)
      |> List.to_tuple()

    start..finish
    |> Enum.map(&digits/1)
  end

  def part1(input) do
    input
    |> Stream.filter(&increasing?/1)
    |> Stream.filter(&double?/1)
    |> Enum.count()
  end

  def part2(input) do
    input
    |> Stream.filter(&increasing?/1)
    |> Stream.filter(&double_strict?/1)
    |> Enum.count()
  end

  defp digits(guess) do
    guess
    |> to_charlist()
    |> Enum.map(&(&1 - ?0))
  end

  defp increasing?(digits) do
    digits
    |> Enum.chunk_every(2, 1, :discard)
    |> Enum.all?(fn [a, b] -> a <= b end)
  end

  defp double?(digits) do
    digits
    |> Stream.chunk_by(& &1)
    |> Enum.any?(&(Enum.count(&1) >= 2))
  end

  defp double_strict?(digits) do
    digits
    |> Stream.chunk_by(& &1)
    |> Enum.any?(&(Enum.count(&1) === 2))
  end
end

input = AoC2019.Day04.parse_input()

input
|> AoC2019.Day04.part1()
|> IO.puts()

input
|> AoC2019.Day04.part2()
|> IO.puts()
